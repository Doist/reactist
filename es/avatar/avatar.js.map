{"version":3,"file":"avatar.js","sources":["../../src/avatar/avatar.tsx"],"sourcesContent":["import * as React from 'react'\n\nimport { getInitials, emailToIndex } from './utils'\n\nimport { getClassNames, ResponsiveProp } from '../utils/responsive-props'\nimport styles from './avatar.module.css'\nimport { Box } from '../box'\nimport type { ObfuscatedClassName } from '../utils/common-types'\n\nconst AVATAR_COLORS = [\n    '#fcc652',\n    '#e9952c',\n    '#e16b2d',\n    '#d84b40',\n    '#e8435a',\n    '#e5198a',\n    '#ad3889',\n    '#86389c',\n    '#a8a8a8',\n    '#98be2f',\n    '#5d9d50',\n    '#5f9f85',\n    '#5bbcb6',\n    '#32a3bf',\n    '#2bafeb',\n    '#2d88c3',\n    '#3863cc',\n    '#5e5e5e',\n]\n\ntype AvatarSize = 'xxs' | 'xs' | 's' | 'm' | 'l' | 'xl' | 'xxl' | 'xxxl'\n\ntype Props = ObfuscatedClassName & {\n    /** @deprecated Please use `exceptionallySetClassName` */\n    className?: string\n    /** @deprecated */\n    colorList?: string[]\n    size?: ResponsiveProp<AvatarSize>\n    avatarUrl?: string\n    user: { name?: string; email: string }\n}\n\nfunction Avatar({\n    user,\n    avatarUrl,\n    size = 'l',\n    className,\n    colorList = AVATAR_COLORS,\n    exceptionallySetClassName,\n    ...props\n}: Props) {\n    const userInitials = getInitials(user.name) || getInitials(user.email)\n    const avatarSize = size ? size : 'l'\n\n    const style = avatarUrl\n        ? {\n              backgroundImage: `url(${avatarUrl})`,\n              textIndent: '-9999px', // hide the initials\n          }\n        : {\n              backgroundColor: colorList[emailToIndex(user.email, colorList.length)],\n          }\n\n    const sizeClassName = getClassNames(styles, 'size', avatarSize)\n\n    return (\n        <Box\n            className={[className, styles.avatar, sizeClassName, exceptionallySetClassName]}\n            style={style}\n            {...props}\n        >\n            {userInitials}\n        </Box>\n    )\n}\nAvatar.displayName = 'Avatar'\n\nexport { Avatar }\n"],"names":["AVATAR_COLORS","Avatar","user","avatarUrl","size","className","colorList","exceptionallySetClassName","props","userInitials","getInitials","name","email","avatarSize","style","backgroundImage","textIndent","backgroundColor","emailToIndex","length","sizeClassName","getClassNames","styles","React","createElement","Box","_objectSpread","avatar","displayName"],"mappings":";;;;;;;;AASA,MAAMA,aAAa,GAAG,CAClB,SADkB,EAElB,SAFkB,EAGlB,SAHkB,EAIlB,SAJkB,EAKlB,SALkB,EAMlB,SANkB,EAOlB,SAPkB,EAQlB,SARkB,EASlB,SATkB,EAUlB,SAVkB,EAWlB,SAXkB,EAYlB,SAZkB,EAalB,SAbkB,EAclB,SAdkB,EAelB,SAfkB,EAgBlB,SAhBkB,EAiBlB,SAjBkB,EAkBlB,SAlBkB,CAAtB,CAAA;;AAiCA,SAASC,MAAT,CAQQ,IAAA,EAAA;EAAA,IARQ;IACZC,IADY;IAEZC,SAFY;AAGZC,IAAAA,IAAI,GAAG,GAHK;IAIZC,SAJY;AAKZC,IAAAA,SAAS,GAAGN,aALA;AAMZO,IAAAA,yBAAAA;GAEI,GAAA,IAAA;AAAA,MADDC,KACC,GAAA,wBAAA,CAAA,IAAA,EAAA,SAAA,CAAA,CAAA;;AACJ,EAAA,MAAMC,YAAY,GAAGC,WAAW,CAACR,IAAI,CAACS,IAAN,CAAX,IAA0BD,WAAW,CAACR,IAAI,CAACU,KAAN,CAA1D,CAAA;AACA,EAAA,MAAMC,UAAU,GAAGT,IAAI,GAAGA,IAAH,GAAU,GAAjC,CAAA;EAEA,MAAMU,KAAK,GAAGX,SAAS,GACjB;IACIY,eAAe,EAAA,MAAA,GAASZ,SAAT,GADnB,GAAA;IAEIa,UAAU,EAAE,SAFhB;;AAAA,GADiB,GAKjB;AACIC,IAAAA,eAAe,EAAEX,SAAS,CAACY,YAAY,CAAChB,IAAI,CAACU,KAAN,EAAaN,SAAS,CAACa,MAAvB,CAAb,CAAA;GANpC,CAAA;EASA,MAAMC,aAAa,GAAGC,aAAa,CAACC,gBAAD,EAAS,MAAT,EAAiBT,UAAjB,CAAnC,CAAA;AAEA,EAAA,oBACIU,KAAC,CAAAC,aAAD,CAACC,GAAD,EAAAC,cAAA,CAAA;IACIrB,SAAS,EAAE,CAACA,SAAD,EAAYiB,gBAAM,CAACK,MAAnB,EAA2BP,aAA3B,EAA0Cb,yBAA1C,CADf;AAEIO,IAAAA,KAAK,EAAEA,KAAAA;GACHN,EAAAA,KAHR,CAKKC,EAAAA,YALL,CADJ,CAAA;AASH,CAAA;;AACDR,MAAM,CAAC2B,WAAP,GAAqB,QAArB;;;;"}