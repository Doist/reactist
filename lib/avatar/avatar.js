"use strict";Object.defineProperty(exports,"__esModule",{value:!0});var e=require("../_virtual/_rollupPluginBabelHelpers.js"),a=require("react"),t=require("./utils.js"),r=require("../utils/responsive-props.js"),s=require("./avatar.module.css.js"),l=require("../box/box.js");function i(e){if(e&&e.__esModule)return e;var a=Object.create(null);return e&&Object.keys(e).forEach((function(t){if("default"!==t){var r=Object.getOwnPropertyDescriptor(e,t);Object.defineProperty(a,t,r.get?r:{enumerable:!0,get:function(){return e[t]}})}})),a.default=e,a}var c=i(a);const o=["user","avatarUrl","size","className","colorList","exceptionallySetClassName"],u=["#fcc652","#e9952c","#e16b2d","#d84b40","#e8435a","#e5198a","#ad3889","#86389c","#a8a8a8","#98be2f","#5d9d50","#5f9f85","#5bbcb6","#32a3bf","#2bafeb","#2d88c3","#3863cc","#5e5e5e"];function n(a){let{user:i,avatarUrl:n,size:b="l",className:d,colorList:f=u,exceptionallySetClassName:p}=a,m=e.objectWithoutProperties(a,o);const v=t.getInitials(i.name)||t.getInitials(i.email),j=b||"l",g=n?{backgroundImage:"url("+n+")",textIndent:"-9999px"}:{backgroundColor:f[t.emailToIndex(i.email,f.length)]},x=r.getClassNames(s.default,"size",j);return c.createElement(l.Box,e.objectSpread2({className:[d,s.default.avatar,x,p],style:g},m),v)}n.displayName="Avatar",exports.Avatar=n;
//# sourceMappingURL=avatar.js.map
