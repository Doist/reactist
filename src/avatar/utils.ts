function getInitials(name?: string) {
    if (!name) {
        return ''
    }

    const seed = name.trim().split(' ')
    const firstInitial = seed[0]
    const lastInitial = seed[seed.length - 1]

    let initials = firstInitial?.[0]
    if (
        firstInitial != null &&
        lastInitial != null &&
        initials != null &&
        // Better readable this way.
        // eslint-disable-next-line @typescript-eslint/prefer-string-starts-ends-with
        firstInitial[0] !== lastInitial[0]
    ) {
        initials += lastInitial[0]
    }
    return initials?.toUpperCase()
}

function emailToIndex(email: string, maxIndex: number) {
    const seed = email.split('@')[0]
    const hash = seed ? seed.charCodeAt(0) + seed.charCodeAt(seed.length - 1) || 0 : 0
    return hash % maxIndex
}

export { getInitials, emailToIndex }
