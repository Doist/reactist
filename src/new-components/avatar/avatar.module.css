:root {
    --reactist-avatar-size-xxsmall: 16px;
    --reactist-avatar-size-xsmall: 20px;
    --reactist-avatar-size-small: 30px;
    --reactist-avatar-size-medium: 32px;
    --reactist-avatar-size-large: 34px;
    --reactist-avatar-size-xlarge: 48px;
    --reactist-avatar-size-xxlarge: 70px;
    --reactist-avatar-size-xxxlarge: 100px;

    --reactist-avatar-size: var(--reactist-avatar-size-large);
}

.avatar {
    flex-shrink: 0;
    background-position: center;
    color: white;
    text-align: center;
    border-radius: 50%;

    width: var(--reactist-avatar-size);
    height: var(--reactist-avatar-size);
    line-height: var(--reactist-avatar-size);
    background-size: var(--reactist-avatar-size);
    font-size: calc(var(--reactist-avatar-size) / 2);
}

.size-xxs {
    --reactist-avatar-size: var(--reactist-avatar-size-xxsmall);
}

.size-xs {
    --reactist-avatar-size: var(--reactist-avatar-size-xsmall);
}

.size-s {
    --reactist-avatar-size: var(--reactist-avatar-size-small);
}

.size-m {
    --reactist-avatar-size: var(--reactist-avatar-size-medium);
}

.size-l {
    --reactist-avatar-size: var(--reactist-avatar-size-large);
}

.size-xl {
    --reactist-avatar-size: var(--reactist-avatar-size-xlarge);
}

.size-xxl {
    --reactist-avatar-size: var(--reactist-avatar-size-xxlarge);
}

.size-xxxl {
    --reactist-avatar-size: var(--reactist-avatar-size-xxxlarge);
}

/* padding-top for tablet */
@media (min-width: 768px /* --reactist-breakpoint-tablet */) {
    .tablet-size-xxs {
        --reactist-avatar-size: var(--reactist-avatar-size-xxsmall);
    }

    .tablet-size-xs {
        --reactist-avatar-size: var(--reactist-avatar-size-xsmall);
    }

    .tablet-size-s {
        --reactist-avatar-size: var(--reactist-avatar-size-small);
    }

    .tablet-size-m {
        --reactist-avatar-size: var(--reactist-avatar-size-medium);
    }

    .tablet-size-l {
        --reactist-avatar-size: var(--reactist-avatar-size-large);
    }

    .tablet-size-xl {
        --reactist-avatar-size: var(--reactist-avatar-size-xlarge);
    }

    .tablet-size-xxl {
        --reactist-avatar-size: var(--reactist-avatar-size-xxlarge);
    }

    .tablet-size-xxxl {
        --reactist-avatar-size: var(--reactist-avatar-size-xxxlarge);
    }
}

/* padding-top for desktop */
@media (min-width: 992px /* --reactist-breakpoint-desktop */) {
    .desktop-size-xxs {
        --reactist-avatar-size: var(--reactist-avatar-size-xxsmall);
    }

    .desktop-size-xs {
        --reactist-avatar-size: var(--reactist-avatar-size-xsmall);
    }

    .desktop-size-s {
        --reactist-avatar-size: var(--reactist-avatar-size-small);
    }

    .desktop-size-m {
        --reactist-avatar-size: var(--reactist-avatar-size-medium);
    }

    .desktop-size-l {
        --reactist-avatar-size: var(--reactist-avatar-size-large);
    }

    .desktop-size-xl {
        --reactist-avatar-size: var(--reactist-avatar-size-xlarge);
    }

    .desktop-size-xxl {
        --reactist-avatar-size: var(--reactist-avatar-size-xxlarge);
    }

    .desktop-size-xxxl {
        --reactist-avatar-size: var(--reactist-avatar-size-xxxlarge);
    }
}
