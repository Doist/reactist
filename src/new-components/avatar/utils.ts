function getInitials(name?: string) {
    if (!name) {
        return ''
    }

    const seed = name.trim().split(' ')
    const firstInitial = seed[0]
    const lastInitial = seed[seed.length - 1]

    let initials = firstInitial[0]
    // Better readable this way.
    // eslint-disable-next-line @typescript-eslint/prefer-string-starts-ends-with
    if (firstInitial[0] !== lastInitial[0]) {
        initials += lastInitial[0]
    }
    return initials.toUpperCase()
}

function emailToIndex(email: string, maxIndex: number) {
    const seed = email.split('@')[0]
    const hash = seed.charCodeAt(0) + seed.charCodeAt(seed.length - 1) || 0
    return hash % maxIndex
}

export { getInitials, emailToIndex }
