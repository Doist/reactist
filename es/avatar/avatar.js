import {
    objectWithoutProperties as _objectWithoutProperties,
    objectSpread2 as _objectSpread2,
} from '../_virtual/_rollupPluginBabelHelpers.js'
import * as React from 'react'
import { getInitials, emailToIndex } from './utils.js'
import { getClassNames } from '../utils/responsive-props.js'
import modules_08f3eeac from './avatar.module.css.js'
import { Box } from '../box/box.js'

const _excluded = [
    'user',
    'avatarUrl',
    'size',
    'className',
    'colorList',
    'exceptionallySetClassName',
]
const AVATAR_COLORS = [
    '#fcc652',
    '#e9952c',
    '#e16b2d',
    '#d84b40',
    '#e8435a',
    '#e5198a',
    '#ad3889',
    '#86389c',
    '#a8a8a8',
    '#98be2f',
    '#5d9d50',
    '#5f9f85',
    '#5bbcb6',
    '#32a3bf',
    '#2bafeb',
    '#2d88c3',
    '#3863cc',
    '#5e5e5e',
]

function Avatar(_ref) {
    let {
            user,
            avatarUrl,
            size = 'l',
            className,
            colorList = AVATAR_COLORS,
            exceptionallySetClassName,
        } = _ref,
        props = _objectWithoutProperties(_ref, _excluded)

    const userInitials = getInitials(user.name) || getInitials(user.email)
    const avatarSize = size ? size : 'l'
    const style = avatarUrl
        ? {
              backgroundImage: 'url(' + avatarUrl + ')',
              textIndent: '-9999px', // hide the initials
          }
        : {
              backgroundColor: colorList[emailToIndex(user.email, colorList.length)],
          }
    const sizeClassName = getClassNames(modules_08f3eeac, 'size', avatarSize)
    return /*#__PURE__*/ React.createElement(
        Box,
        _objectSpread2(
            {
                className: [
                    className,
                    modules_08f3eeac.avatar,
                    sizeClassName,
                    exceptionallySetClassName,
                ],
                style: style,
            },
            props,
        ),
        userInitials,
    )
}

Avatar.displayName = 'Avatar'

export { Avatar }
//# sourceMappingURL=avatar.js.map
