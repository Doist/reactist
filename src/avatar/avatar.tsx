import { Box } from '../box'
import { getClassNames } from '../utils/responsive-props'

import { emailToIndex, getInitials } from './utils'

import styles from './avatar.module.css'

import type { ObfuscatedClassName } from '../utils/common-types'
import type { ResponsiveProp } from '../utils/responsive-props'

const AVATAR_COLORS = [
    '#fcc652',
    '#e9952c',
    '#e16b2d',
    '#d84b40',
    '#e8435a',
    '#e5198a',
    '#ad3889',
    '#86389c',
    '#a8a8a8',
    '#98be2f',
    '#5d9d50',
    '#5f9f85',
    '#5bbcb6',
    '#32a3bf',
    '#2bafeb',
    '#2d88c3',
    '#3863cc',
    '#5e5e5e',
]

type AvatarSize = 'xxs' | 'xs' | 's' | 'm' | 'l' | 'xl' | 'xxl' | 'xxxl'

type Props = ObfuscatedClassName & {
    /** @deprecated Please use `exceptionallySetClassName` */
    className?: string
    /** @deprecated */
    colorList?: string[]
    size?: ResponsiveProp<AvatarSize>
    avatarUrl?: string
    user: { name?: string; email: string }
}

function Avatar({
    user,
    avatarUrl,
    size = 'l',
    className,
    colorList = AVATAR_COLORS,
    exceptionallySetClassName,
    ...props
}: Props) {
    const userInitials = getInitials(user.name) || getInitials(user.email)
    const avatarSize = size ? size : 'l'

    const style = avatarUrl
        ? {
              backgroundImage: `url(${avatarUrl})`,
              textIndent: '-9999px', // hide the initials
          }
        : {
              backgroundColor: colorList[emailToIndex(user.email, colorList.length)],
          }

    const sizeClassName = getClassNames(styles, 'size', avatarSize)

    return (
        <Box
            className={[className, styles.avatar, sizeClassName, exceptionallySetClassName]}
            style={style}
            {...props}
        >
            {userInitials}
        </Box>
    )
}
Avatar.displayName = 'Avatar'

export { Avatar }
