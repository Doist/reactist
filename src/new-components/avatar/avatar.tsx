import React from 'react'

import { getInitials, emailToIndex } from './utils'

import { getClassNames, ResponsiveProp } from '../responsive-props'
import styles from './avatar.module.css'
import { Box } from '../box'

const AVATAR_COLORS = [
    '#fcc652',
    '#e9952c',
    '#e16b2d',
    '#d84b40',
    '#e8435a',
    '#e5198a',
    '#ad3889',
    '#86389c',
    '#a8a8a8',
    '#98be2f',
    '#5d9d50',
    '#5f9f85',
    '#5bbcb6',
    '#32a3bf',
    '#2bafeb',
    '#2d88c3',
    '#3863cc',
    '#5e5e5e',
]

type AvatarSize = 'xxs' | 'xs' | 's' | 'm' | 'l' | 'xl' | 'xxl' | 'xxxl'

type Props = {
    /** @deprecated Please use `exceptionallySetClassName` */
    className?: string
    exceptionallySetClassName?: string
    colorList?: string[]
    size?: ResponsiveProp<AvatarSize>
    avatarUrl?: string
    user: { name?: string; email: string }
}

function Avatar({ user, avatarUrl, size = 'l', className, exceptionallySetClassName }: Props) {
    const userInitials = getInitials(user.name) || getInitials(user.email)
    const avatarSize = size ? size : 'l'

    const style = avatarUrl
        ? {
              backgroundImage: `url(${avatarUrl})`,
              textIndent: '-9999px', // hide the initials
          }
        : {
              backgroundColor: AVATAR_COLORS[emailToIndex(user.email, AVATAR_COLORS.length)],
          }

    const sizeClassName = getClassNames(styles, 'size', avatarSize)

    return (
        <Box
            className={[className, styles.avatar, sizeClassName, exceptionallySetClassName]}
            style={style}
        >
            {userInitials}
        </Box>
    )
}
Avatar.displayName = 'Avatar'

export { Avatar }
