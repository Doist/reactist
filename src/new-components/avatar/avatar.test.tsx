import React from 'react'
import { shallow } from 'enzyme'

import { Avatar } from './avatar'

describe('Avatar', () => {
    it('renders a background image when avatarUrl is supplied', () => {
        const avatar = shallow(getAvatar({ avatarUrl: 'https://foo.bar/com.png' }))
        expect(avatar).toMatchSnapshot()
    })

    it('renders initials of user name when avatarUrl is not supplied', () => {
        const avatar = shallow(getAvatar())
        expect(avatar).toMatchSnapshot()
        expect(avatar.text()).toBe('HM')
    })

    it('renders initials on custom color', () => {
        const avatar = shallow(getAvatar({ colorList: ['red', 'green', '#0000FF'] }))
        expect(avatar).toMatchSnapshot()
        expect(avatar.text()).toBe('HM')
    })

    it('renders initials of user email when avatarUrl is not supplied', () => {
        const avatar = shallow(getAvatar({ user: { email: 'henning@doist.com' } }))
        expect(avatar).toMatchSnapshot()
        expect(avatar.text()).toBe('H')
    })

    it('supports responsive values', () => {
        const avatar = shallow(
            getAvatar({
                size: {
                    mobile: 's',
                    desktop: 'xl',
                    tablet: 'xxl',
                },
            }),
        )
        expect(avatar.hasClass('size-s')).toBe(true)
        expect(avatar.hasClass('desktop-size-xl')).toBe(true)
        expect(avatar.hasClass('tablet-size-xxl')).toBe(true)
    })

    // Helpers ================================================================
    const getAvatar = (
        props?: Omit<React.ComponentProps<typeof Avatar>, 'user'> & {
            user?: { name?: string; email: string }
        },
    ) => <Avatar user={{ name: 'Henning Mus', email: 'henning@doist.com' }} size="xl" {...props} />
})
