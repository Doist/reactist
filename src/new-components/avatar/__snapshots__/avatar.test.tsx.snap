// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Avatar renders a background image when avatarUrl is supplied 1`] = `
<div
  class="avatar size-xl box"
  data-testid="avatar"
  style="background-image: url(https://foo.bar/com.png); text-indent: -9999px;"
>
  HM
</div>
`;

exports[`Avatar renders initials of user email when avatarUrl is not supplied 1`] = `
<div
  class="avatar size-xl box"
  data-testid="avatar"
  style="background-color: rgb(152, 190, 47);"
>
  H
</div>
`;

exports[`Avatar renders initials of user name when avatarUrl is not supplied 1`] = `
<div
  class="avatar size-xl box"
  data-testid="avatar"
  style="background-color: rgb(152, 190, 47);"
>
  HM
</div>
`;

exports[`Avatar renders initials on custom color 1`] = `
<div
  class="avatar size-xl box"
  data-testid="avatar"
  style="background-color: red;"
>
  HM
</div>
`;
